File: backend/app/schemas/knowledge.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field

from app.models.enums import (
    EntityKind,
    ManuscriptLinkRole,
    RelationshipKind,
)
from app.schemas._common import TimestampedRead


# --- entity ---------------------------------------------------------------


class KnowledgeEntityCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    slug: Optional[str] = Field(default=None, max_length=220)
    kind: EntityKind = EntityKind.THEME
    description: Optional[str] = None
    extras: Optional[str] = None  # JSON-encoded text


class KnowledgeEntityUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    slug: Optional[str] = Field(default=None, max_length=220)
    kind: Optional[EntityKind] = None
    description: Optional[str] = None
    extras: Optional[str] = None


class KnowledgeEntityRead(TimestampedRead):
    name: str
    slug: str
    kind: EntityKind
    description: Optional[str]
    extras: Optional[str]


# --- relationship ---------------------------------------------------------


class KnowledgeRelationshipCreate(BaseModel):
    source_id: str
    target_id: str
    kind: RelationshipKind = RelationshipKind.RELATED_TO
    weight: Optional[float] = Field(default=None, ge=0, le=1)
    description: Optional[str] = None


class KnowledgeRelationshipUpdate(BaseModel):
    kind: Optional[RelationshipKind] = None
    weight: Optional[float] = Field(default=None, ge=0, le=1)
    description: Optional[str] = None


class KnowledgeRelationshipRead(TimestampedRead):
    source_id: str
    target_id: str
    kind: RelationshipKind
    weight: Optional[float]
    description: Optional[str]


class KnowledgeRelationshipDetail(KnowledgeRelationshipRead):
    """Edge plus the names of its endpoints for direct rendering."""

    source_name: Optional[str] = None
    source_kind: Optional[EntityKind] = None
    target_name: Optional[str] = None
    target_kind: Optional[EntityKind] = None


# --- manuscript link ------------------------------------------------------


class ManuscriptEntityLinkCreate(BaseModel):
    manuscript_id: str
    entity_id: str
    role: ManuscriptLinkRole = ManuscriptLinkRole.TAGGED
    relevance: Optional[float] = Field(default=None, ge=0, le=1)
    notes: Optional[str] = None


class ManuscriptEntityLinkUpdate(BaseModel):
    role: Optional[ManuscriptLinkRole] = None
    relevance: Optional[float] = Field(default=None, ge=0, le=1)
    notes: Optional[str] = None


class ManuscriptEntityLinkRead(TimestampedRead):
    manuscript_id: str
    entity_id: str
    role: ManuscriptLinkRole
    relevance: Optional[float]
    notes: Optional[str]
    entity_name: Optional[str] = None
    entity_kind: Optional[EntityKind] = None
    entity_slug: Optional[str] = None


# --- neighborhood ---------------------------------------------------------


class NeighborhoodNode(BaseModel):
    id: str
    name: str
    slug: str
    kind: EntityKind
    distance: int  # 0 for root


class NeighborhoodEdge(BaseModel):
    id: str
    source_id: str
    target_id: str
    kind: RelationshipKind
    weight: Optional[float] = None
    description: Optional[str] = None


class NeighborhoodResult(BaseModel):
    root_id: str
    depth: int
    nodes: list[NeighborhoodNode]
    edges: list[NeighborhoodEdge]

File: frontend/src/types/knowledge.ts
export type EntityKind =
  | 'character'
  | 'place'
  | 'theme'
  | 'motif'
  | 'organization'
  | 'work'
  | 'person'
  | 'period'
  | 'other';

export const ENTITY_KINDS: EntityKind[] = [
  'theme',
  'motif',
  'character',
  'person',
  'place',
  'period',
  'work',
  'organization',
  'other',
];

export const ENTITY_KIND_LABEL: Record<EntityKind, string> = {
  character: 'Character',
  place: 'Place',
  theme: 'Theme',
  motif: 'Motif',
  organization: 'Organization',
  work: 'Work',
  person: 'Person',
  period: 'Period',
  other: 'Other',
};

export type RelationshipKind =
  | 'related_to'
  | 'influences'
  | 'descends_from'
  | 'contrasts_with'
  | 'inhabits'
  | 'authored'
  | 'part_of'
  | 'sibling_of'
  | 'mentor_of'
  | 'adapts'
  | 'other';

export const RELATIONSHIP_KIND_LABEL: Record<RelationshipKind, string> = {
  related_to: 'related to',
  influences: 'influences',
  descends_from: 'descends from',
  contrasts_with: 'contrasts with',
  inhabits: 'inhabits',
  authored: 'authored',
  part_of: 'part of',
  sibling_of: 'sibling of',
  mentor_of: 'mentor of',
  adapts: 'adapts',
  other: 'other',
};

export type ManuscriptLinkRole =
  | 'tagged'
  | 'features'
  | 'references'
  | 'set_in'
  | 'derived_from'
  | 'other';

export const MANUSCRIPT_LINK_ROLES: ManuscriptLinkRole[] = [
  'tagged',
  'features',
  'references',
  'set_in',
  'derived_from',
  'other',
];

export const MANUSCRIPT_LINK_ROLE_LABEL: Record<ManuscriptLinkRole, string> = {
  tagged: 'Tagged',
  features: 'Features',
  references: 'References',
  set_in: 'Set in',
  derived_from: 'Derived from',
  other: 'Other',
};

export interface KnowledgeEntity {
  id: string;
  created_at: string;
  updated_at: string;
  name: string;
  slug: string;
  kind: EntityKind;
  description: string | null;
  extras: string | null;
}

export interface KnowledgeRelationship {
  id: string;
  created_at: string;
  updated_at: string;
  source_id: string;
  target_id: string;
  kind: RelationshipKind;
  weight: number | null;
  description: string | null;
}

export interface KnowledgeRelationshipDetail extends KnowledgeRelationship {
  source_name: string | null;
  source_kind: EntityKind | null;
  target_name: string | null;
  target_kind: EntityKind | null;
}

export interface ManuscriptEntityLink {
  id: string;
  created_at: string;
  updated_at: string;
  manuscript_id: string;
  entity_id: string;
  role: ManuscriptLinkRole;
  relevance: number | null;
  notes: string | null;
  entity_name: string | null;
  entity_kind: EntityKind | null;
  entity_slug: string | null;
}

export interface NeighborhoodNode {
  id: string;
  name: string;
  slug: string;
  kind: EntityKind;
  distance: number;
}

export interface NeighborhoodEdge {
  id: string;
  source_id: string;
  target_id: string;
  kind: RelationshipKind;
  weight: number | null;
  description: string | null;
}

export interface NeighborhoodResult {
  root_id: string;
  depth: number;
  nodes: NeighborhoodNode[];
  edges: NeighborhoodEdge[];
}
